includes/core/README.md

```diff
@@ -1,3 +1,105 @@
+4. __Extensions__ are those use-cases, which are linked to WooCommerce Payments, but not build on top of/within the plugin. Those should use *WordPress actions and filters* to integrate with the gateway, and connect with the plugin through the classes, contained within the plugin core.
+
+## Support and backwards compatibility
+
+The core is developed, and will be maintained with backwards compatibility in mind.
+
+⚠️ This means that everything outside of this directory should be considered internal for the plugin, and may be changed without consideration for notice.
+
+To ensure stability and resilience, please avoid using non-core classes and methods within your extensions. The plugin is under active development, and if you need something, which is not present in this directory, please [open an issue](https://github.com/Automattic/woocommerce-payments/issues/new/). We will always consider the best way to expose the needed functionality.
+
+## Available functionality
+
+WooCommerce Payments Core is under active development, and currently includes the APIs, listed below.
+
+### Gateway Mode
+
+This is a singular `WCPay\Core\Mode` object, accessible through `WC_Payments::mode()`. It allows you to check whether the gateway is in any of the listed modes, as well as to force it into one of them:
+
+| Mode     | Check      | Setter   | Description                           |
+|----------|------------|----------|---------------------------------------|
+| __Live__ | `is_live()`| `live()` | Standard working mode of the gateway. |
+| __Test__ | `is_test()`| `test()` | Allows test mode on live accounts.    |
+| __Dev__  | `is_dev()` | `dev()`  | Allows test accounts to be created.   |
+
+During initialization, the following logic is used:
+
+1. Dev mode would be entered if:
+    - Either [WordPress's environment type](https://developer.wordpress.org/reference/functions/wp_get_environment_type/#description) is either `development` or 		`staging`.
+    - or `WCPAY_DEV_MODE` is defined and set to boolean true.
+2. Test mode is entered if:
+    - Either Dev mode is already enabled.
+    - or the gateway's test mode setting is on.
+3. If the gateway is neither in dev or test mode, live mode is entered.
+
+To alter this behavior, you can use the `wcpay_dev_mode` and `wcpay_test_mode` filters, for example:
+
+```php
+add_filter( 'wcpay_dev_mode', '__return_true' );
+```
+
+This is done on the `init` action.
+
+> 🎗️ The object is not accessible before the `init` WordPress action.
+
+Once the mode has been initialized, the gateway will work in that mode. The setters (from the table above) may be used to switch the working mode during runtime, but it should be done with caution, as other functionality may already be initialized in the main mode, hence it is recommended to switch back once done.
+
+__Checking the current mode:__
```

Comment:
This is intentional. As mentioned above,

> Once the mode has been initialized, the gateway will work in that mode. The setters (from the table above) may be used to switch the working mode during runtime, but __it should be done with caution__, as other functionality may already be initialized in the main mode, hence it is recommended to switch back once done.

Except for tests, the mode should mostly not be changed once initialized, as it can lead to quirky behavior. The idea of creating the class is exactly to make sure that the filter is only executed once, instead of every time `is_in_dev_mode()` or `is_in_live_mode()` is called. Basically I want to create a small barrier here.

Reply:
Hmm, interesting thought. Makes sense!